cerealize/cerealize.h

```diff
@@ -225,37 +265,112 @@ class GenericCerealFileParser {
-    try {
+    if (fi_.peek() != std::char_traits<char>::eof()) {
       std::unique_ptr<T_ENTRY> entry;
       si_(entry);
       bricks::rtti::RuntimeTupleDispatcher<typename T_PROCESSOR::BASE_TYPE,
                                            typename T_PROCESSOR::DERIVED_TYPE_LIST>::DispatchCall(*entry.get(),
                                                                                                   processor);
       return true;
-    } catch (cereal::Exception&) {
-      // TODO(dkorolev): Should check whether we have reached the end of the file here, otherwise
-      // distinguishing between EOF vs. de-cereal-ization issue in the middle of the file will be
-      // pain in the ass.
+    } else {
       return false;
     }
   }
 
  private:
-  GenericCerealFileParser() = delete;
-  GenericCerealFileParser(const GenericCerealFileParser&) = delete;
-  void operator=(const GenericCerealFileParser&) = delete;
-  GenericCerealFileParser(GenericCerealFileParser&&) = delete;
-  void operator=(GenericCerealFileParser&&) = delete;
+  CerealBinaryFileParser() = delete;
+  CerealBinaryFileParser(const CerealBinaryFileParser&) = delete;
+  void operator=(const CerealBinaryFileParser&) = delete;
+  CerealBinaryFileParser(CerealBinaryFileParser&&) = delete;
+  void operator=(CerealBinaryFileParser&&) = delete;
 
   std::ifstream fi_;
-  typename CerealStreamType<T_CEREAL_FORMAT>::Input si_;
+  cereal::BinaryInputArchive si_;
 };
+
+
+// `CerealJSONFileParser` de-cereal-izes records from JSON file given their type 
+// and passes them over to `T_PROCESSOR`. Each line in the file expected to be a
+// full JSON record for one entry.
 template <typename T_ENTRY>
-using CerealFileParser = GenericCerealFileParser<T_ENTRY, CerealFormat::Default>;
+class CerealJSONFileParser {
+ public:
+  explicit CerealJSONFileParser(const std::string& filename) : fi_(filename) {}
+
+  // `Next` calls `T_PROCESSOR::operator()(const T_ENTRY&)` for the next entry, or returns false.
+  template <typename T_PROCESSOR>
+  bool Next(T_PROCESSOR&& processor) {
+    std::unique_ptr<T_ENTRY> entry = GetNextEntry();
+    if (entry) {
+      processor(*entry.get());
+      return true;
+    } else {
+      return false;
+    }
+  }
+
+  // `NextWithDispatching` calls `T_PROCESSOR::operator()(const T_ACTUAL_ENTRY_TYPE&)`
+  // for the next entry, or returns false.
+  // Note that the actual entry type is being used in the calling method signature.
+  // This also makes it implausible to pass in a lambda here, thus the parameter is only passed by reference.
+  // In order for RTTI dispatching to work, class T_PROCESSOR should define two type:
+  // 1) BASE_TYPE: The type of the base entry to de-serialize from the stream, and
+  // 2) DERIVED_TYPE_LIST: An std::tuple<TYPE1, TYPE2, TYPE3, ...> of all the types that have to be matched.
+  template <typename T_PROCESSOR>
+  bool NextWithDispatching(T_PROCESSOR& processor) {
+    std::unique_ptr<T_ENTRY> entry = GetNextEntry();
```

Comment:
Nit: Parentheses initialization syntax would read better than assignment initialization syntax here. The way it's written invites adding `std::move()`. :-)